sqlserver/datadog_checks/sqlserver/sqlserver.py

```diff
@@ -725,6 +729,33 @@ def _check_connections_by_use_db(self):
                 # Switch DB back to MASTER
                 cursor.execute(SWITCH_DB_STATEMENT.format(self.connection.DEFAULT_DATABASE))
 
+    def get_databases(self):
+        engine_edition = self.static_info_cache.get(STATIC_INFO_ENGINE_EDITION)
+        if not is_azure_sql_database(engine_edition):
+            db_names = [d.name for d in self.databases] or [
+                self.instance.get('database', self.connection.DEFAULT_DATABASE)
+            ]
+        else:
+            db_names = [self.instance.get('database', self.connection.DEFAULT_DATABASE)]
+        return db_names
+
+    def do_for_databases(self, action, databases):
+        engine_edition = self.static_info_cache.get(STATIC_INFO_ENGINE_EDITION)
+        with self.connection.open_managed_default_connection():
+            with self.connection.get_managed_cursor() as cursor:
+                for db in databases:
+                    try:
+                        if not is_azure_sql_database(engine_edition):
+                            cursor.execute(SWITCH_DB_STATEMENT.format(db))
+                        stop = action(cursor, db)
+                        if stop:
+                            break
```

Comment:
I'm a bit torn on this. On the one hand, a helper function that operates across all databases is nice. On the other hand, I find this control flow a bit strange. It requires the `action` function to know about the loop outside of it. It feels a bit weird, but I can't put my finger on exactly why. I could also see the `false` return value being replaced with a `StopIteration` exception to make it a bit more clear, but that doesn't get at the root of my unease about exposing the outer control flow to an individual function. 

Seems like a good opportunity for the Agent folks to have a good technical discussion about reusable code patterns :)

Reply:
Actually I think my suggestion here might help: https://github.com/DataDog/integrations-core/pull/17258#discussion_r1593152878

Essentially, the inner function would not know about the control flow, but would know to skip DBs if max count is exceeded. So this function would return something like this:

```
>>> errors = do_for_databases(...)
>>> pprint.pprint(errors)
[
    {"master": Exception("User: datadog does not have permission to the master database")}, 
    {"tempdb": TimeoutError("Query timed out")},
    {"mydb0": None},
    {"mydb1": None},
    {"mydb2": DbSkipError("max results exceeded")},
    {"mydb3": DbSkipError("max results exceeded")},
    {"mydb4": DbSkipError("max results exceeded")},
    ...
]
```

Reply:
I did the modif pls check it out. 
At the end if we see that there are more problems I can also do a loop ... inside of a loop and see later if I can introduce do_for_databases in a separate refactoring PR.